[Test] Create TestCalcJoinSingleAssetTokensIn test case to fall outside of the error tolerance

stackman27 opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
v    If your PR doesn't close an issue, that's OK!  Just remove the Closes: #XXX line!
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #1773

## What is the purpose of the change
Check if the `tokenshares` fall outside of the error tolerance

- [x] fix err tolerance check function to accommodate  error 
- [x] test for single token in that falls outside error tolerance 
- [x]  test for multi tokens in that falls outside the error tolerance
- [x]  Add tolerance check for `TestCalcJoinPoolShares` 

## Brief Changelog
n/a

## Testing and Verifying
Add extra test to `TestCalcJoinSingleAssetTokensIn`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not applicable)

mattverse commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -820,11 +854,12 @@ func TestCalcJoinSingleAssetTokensIn(t *testing.T) {
+			swapFee:              sdk.MustNewDecFromStr("0"),
+			poolAssets:           oneTrillionEvenPoolAssets,
+			tokensIn:             sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
+			expectShares:         sdk.NewInt(2_499_999_968_750), // ACTUAL: 249_999_996_8800
```
```suggestion
			expectShares:         sdk.NewInt(2_499_999_968_750), // ACTUAL: 249_999_996_880
```

mattverse commented on x/gamm/pool-models/balancer/util_test.go:
```diff
@@ -39,19 +39,19 @@ func createTestContext(t *testing.T) sdk.Context {
-		errTolerance.Compare(expectedShares, actualShares),
-		fmt.Sprintf("expectedShares: %s, actualShares: %s", expectedShares.String(), actualShares.String()))
+	if errTolerance.Compare(expectedShares, actualShares) != 0 {
+		return errors.New("shares doesnot fall within the tolerance range")
```
```suggestion
		return errors.New("shares does not fall within the tolerance range")
```

mattverse commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -939,24 +977,51 @@ func TestCalcJoinSingleAssetTokensIn(t *testing.T) {
+			expectShares:         sdk.NewInt(2_072_912_400_000_000 + 1_624_999_900_000), // ACTUAL: 2_074_537_403_302_024
+			expectToleranceError: false, 
+		},
+		{
+			// |a - b| / min(a, b) <= allowedErrRatio
+			// a = expected shares, b = actual shares
+			// 0.04000000047920000 = abs(2_599_999_968_750-2_499_999_968_800)/2_499_999_968_800
+			// 	Full solution: https://www.wolframalpha.com/input?i=abs%282599999968750-2499999968800%29%2F2499999968800
+			name: "Error Tolerance: One token in, expected > actual",
+			swapFee:              sdk.MustNewDecFromStr("0"),
+			poolAssets:           oneTrillionEvenPoolAssets,
+			tokensIn:             sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
+			expectShares:         sdk.NewInt(2_599_999_968_750),// ACTUAL: 2_499_999_968_800
```
wdyt about adding comments for the calculations on why these are outside the range of balancer module's `allowedErrRatio`?

stackman27 replied:
I can do that

p0mvn commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -80,6 +81,7 @@ var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
 		poolAssets:   oneTrillionEvenPoolAssets,
 		tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
 		expectShares: sdk.NewInt(2_499_999_968_750),
+		expectToleranceError: false,
```
No need to specify `expectToleranceError: false,` since non-initialized bool is equivalent to false in Go

p0mvn commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -101,6 +103,7 @@ var calcSingleAssetJoinTestCases = []calcJoinSharesTestCase{
 		poolAssets:   oneTrillionEvenPoolAssets,
 		tokensIn:     sdk.NewCoins(sdk.NewInt64Coin("uosmo", 50_000)),
 		expectShares: sdk.NewInt(2_487_500_000_000),
+		expectToleranceError: false, 
```
ditto and for all other instances of `expectToleranceError: false,`

p0mvn commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -603,8 +626,13 @@ func (suite *KeeperTestSuite) TestCalcJoinPoolShares() {
-					assertExpectedSharesErrRatio(t, tc.expectShares, shares)
 					assertExpectedLiquidity(t, tc.tokensIn, liquidity)
+					err = assertExpectedSharesErrRatio(t, tc.expectShares, shares)
+					if tc.expectToleranceError {
```
I don't see a single test case where `expectToleranceError` is true, do we need to have this conditional?

p0mvn commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -786,7 +814,12 @@ func TestCalcSingleAssetJoin(t *testing.T) {
 				require.NoError(t, err)
-				assertExpectedSharesErrRatio(t, tc.expectShares, shares)
+				err = assertExpectedSharesErrRatio(t, tc.expectShares, shares)
+				if tc.expectToleranceError {
```
I don't see a single test case where `expectToleranceError` is true, do we need to have this conditional?

stackman27 replied:
i have added two tests so far; 
1. https://github.com/osmosis-labs/osmosis/pull/1822/files#diff-d90d7fb125be0247c48c167de4adedf2ddbdc76b1ad06bc243d329bc8dbff76dR985
2. https://github.com/osmosis-labs/osmosis/pull/1822/files#diff-d90d7fb125be0247c48c167de4adedf2ddbdc76b1ad06bc243d329bc8dbff76dR997

p0mvn replied:
But these are defined in a different test fixture - `TestCalcJoinSingleAssetTokensIn`, are they not?

The test cases for this particular fixture (`TestCalcJoinPoolShares`) seem to be unmodified

stackman27 replied:
aaah i see! I thought i was only creating tolerance check tests for `TestCalcJoinSingleAssetTokensIn` function. I can create tests for `TestCalcJoinPoolShares` too

p0mvn commented on x/gamm/pool-models/balancer/amm_joinpool_test.go:
```diff
@@ -798,13 +831,14 @@ func TestCalcSingleAssetJoin(t *testing.T) {
-		expectShares   sdk.Int
-		expectLiqudity sdk.Coins
-		expErr         error
+		name                 string
```
can we make this be `calcJoinSharesTestCase`? Seems that most of the fields are the same

p0mvn reviewed (commented):
Nice work so far. Haven't checked the calculations in detail yet but had some comments. Please take a look

stackman27 commented:
@p0mvn @mattverse Thank you for the review! =) 

Addressed all your comments [6d69ca4](https://github.com/osmosis-labs/osmosis/pull/1822/commits/6d69ca4c566202f657de54bd7bdf155dce1caec0)

mattverse reviewed (approved):
Nice work on the vecotrs here! LGTM!

p0mvn commented on x/gamm/pool-models/balancer/pool_test.go:
```diff
@@ -354,6 +351,42 @@ func TestCalcJoinSingleAssetTokensIn(t *testing.T) {
+		{
+			// Compares expected shares and actual shares to make sure the multiplicativeTolerance ratio below is maintained
+			// |a - b| / min(a, b) <= allowedErrRatio
+			// 0.04000000047920000 = abs(2_599_999_968_750-2_499_999_968_800)/2_499_999_968_800
```
Identifying the test cases is an important step toward fixing these issues. You've done a great job at this! Now, we should understand the sources of these errors so that we can fix them.

On that note, I'm wondering: 

- How did you choose these values so that they are outside of the error tolerance? In other words, for each test case, what choice of inputs causes the delta to be large? 

Could you please document this directly in the comments? I think it would be incredibly helpful so that we can come back to it later and know what causes these issues.

p0mvn commented on x/gamm/pool-models/balancer/pool_suite_test.go:
```diff
@@ -51,14 +51,15 @@ var (
+	expectLiq            sdk.Coins
+	expectPanic          bool
+	expErr               error
+	expectToleranceError bool
```
```suggestion
        // There are known cases where the actual shares might fall outside of the
        // tolerable range from the expected shares. Each such test case documents
        // the reason for the fallout. 
	expectToleranceError bool
```

p0mvn reviewed (commented):
Great work! I'm currently not sure what inputs cause the delta to be large. Can we document that for each test case, please?

p0mvn commented on x/gamm/pool-models/balancer/pool_suite_test.go:
```diff
@@ -650,6 +654,56 @@ func (suite *KeeperTestSuite) TestCalcJoinPoolShares() {
+				sdk.NewInt64Coin("uosmo", 50_000),
+				sdk.NewInt64Coin("uatom", 75_000),
+			),
+			expectShares:         sdk.NewInt(2.5e12 + 1350000000000), // ACTUAL: 6_243_749_992_287
```
Also, is there a Wolfram calculation for `expectShares`? How was this value selected?

stackman27 replied:
I thought about this previously as well, but if i calculate expShares(through our logic)  and insert the right value, it wouldn't cause tolerance err right? So i purposely tried to mess around the numbers

p0mvn replied:
So our goal is to find out if it is possible to find test cases with valid inputs / settings that cause us to fall outside of the error tolerance.

That would mean having a valid calculation for `expectShares` , similar to previous test cases that, at the same time, fall outside of the error tolerance.

How can we claim that calculated `shares` fall outside of the error tolerance from `expectShares` when the `expectShares are randomly selected, right?

So, for this to be accepted, I would like to know how `expectShares` are estimated. Preferably, with a Wolfram link similar to previous test cases.

stackman27 replied:
hmm i tried lots of scenarios and i wasn't able to generate scenarios to get values that fall outside the error tolerance. In face the larger the `tokensIn` the better the bound became. Maybe we might wanna make the error tolerance higher?  

This is the largest value i tried; 
```
 Desc: one token in - equal weights with zero swap fee
 P_issued = P_supply * ((1 + (A_t / B_t))^W_t - 1)
 4_880_884_817_015_154_699 =  1e20 * (( 1 + (100000000000 / 1e12))^0.5 - 1) 

|a - b| / min(a, b) <= allowedErrRatio
0.000000000000000000204 = abs(4_880_884_817_015_154_699-4_880_884_817_015_154_700)/4_880_884_817_015_154_699

0.000000000000000000204 <= 0.0000001 (WITHIN ACCEPTED RATIO)
```

stackman27 replied:
cc @p0mvn

p0mvn replied:
That makes sense, thanks for looking deep into this.

If that's the case, let's do the following:
- add any test cases that you've tried and think make sense to add
- make comments consistent with the rest of the test cases
- remove prints and `expectToleranceError bool`

Then, we can close this :+1:  The conclusion is that we could not find a case to fall outside of the tolerance
